src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,200 @@
+
+            DataTable toReturn = new DataTable(ProviderGroupColumnName) { Locale = CultureInfo.InvariantCulture };
+            toReturn.Columns.AddRange(new[] { nameColumn, descriptionColumn, invariantNameColumn, assemblyQualifiedNameColumn });
+            toReturn.PrimaryKey = new[] { invariantNameColumn };
+            foreach(var kvp in _registeredFactories)
+            {
+                DataRow newRow = toReturn.NewRow();
+                newRow[InvariantNameColumnName] = kvp.Key;
+                newRow[AssemblyQualifiedNameColumnName] = kvp.Value.FactoryTypeAssemblyQualifiedName;
+                newRow[NameColumnName] = string.Empty;
+                newRow[DescriptionColumnName] = string.Empty;
+                toReturn.AddRow(newRow);
+            }
+            return toReturn;
+        }
+
+        public static IEnumerable<string> GetProviderInvariantNames()
+        {
+            return _registeredFactories.Keys.ToList();
+        }
+
+        public static void RegisterFactory(string providerInvariantName, string factoryTypeAssemblyQualifiedName)
+        {
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            ADP.CheckArgumentLength(factoryTypeAssemblyQualifiedName, nameof(factoryTypeAssemblyQualifiedName));
+            
+            // this method performs a deferred registration: the type name specified is checked when the factory is requested for the first time. 
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factoryTypeAssemblyQualifiedName, null);
+        }
+        
+        public static void RegisterFactory(string providerInvariantName, Type providerFactoryClass)
+        {
+            RegisterFactory(providerInvariantName, GetFactoryInstance(providerFactoryClass));
+        }
+
+        public static void RegisterFactory(string providerInvariantName, DbProviderFactory factory)
+        {
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            ADP.CheckArgumentNull(factory, nameof(factory));
+
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factory.GetType().AssemblyQualifiedName, factory);
+        }
+        
+        public static bool UnregisterFactory(string providerInvariantName)
+        {
+            return !string.IsNullOrWhiteSpace(providerInvariantName) && _registeredFactories.TryRemove(providerInvariantName, out ProviderRegistration providerRegistration);
+        }
+        
+        private static DbProviderFactory GetFactory(string providerInvariantName, bool throwOnError)
+        {
+            if (throwOnError)
+            {
+                ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            }
+            else
+            {
+                if (string.IsNullOrWhiteSpace(providerInvariantName))
```

Comment:
Nit: collapse into `else if`?

Reply:
I hate 'else if' as it has an 'else' block without curly braces, which I find personally not a good choice, every block should have curly braces, even if it's just 1 statement.